proposals/p3833.md

```diff
@@ -0,0 +1,211 @@
+behavior in the toolchain. This means starting with that implementation doesn't
+result in sunk cost.
+
+### Example: overloaded operators
+
+> Note: this example isn't trying to capture all the nuance of the current
+> design or suggest any change to the design. It is merely trying to provide an
+> illustration of how we might follow the proposal in practice.
+
+As an example to help illustrate this, how should the toolchain model overloaded
+operators? This proposal suggests that _initially_, the implementation should
+aim to fully model the rewrite-based dispatch through interfaces in the prelude.
+That is, each use of `x + y` should turn into roughly equivalent SemIR as would
+be used to model the rewritten semantics of `x.(Core.AddWith(typeof(y)).Op)(y)`.
+This in turn would dispatch to an exact-type implementation which would provide
+any implicit conversions, and so-on.
+
+To be precise, the expectation is that the SemIR for `x + y` should as a
+consequence model all of:
+
+-   Looking up the `Core.AddWith` interface.
+-   Passing the type of `y` as a parameter to the interface (but likely using
+    that type directly rather than forming a literal `typeof(y)` kind of
+    expression, see below).
+-   Looking up the `Op` method.
+-   Calling that `Op` method on `x` with `y` as its parameter just as we would
+    model any call to `object.(Interface.Method)(y)`.
+
+However, specific things that would still be reasonable without deviating from
+the full semantic modeling here:
+
+-   Not representing a user-written `typeof` style expression to compute the
+    type, or duplicating the `y` expression.
+-   Caching and re-using SemIR for immutable things like successfully resolved
+    name lookup. Concretely, re-using a single cached lookup of `Core.AddWith`
+    would still provide the full model. And similarly, re-using an already
+    computed parameterization such as `Core.AddWith(i32)` repeatedly, and the
+    lookup of `Op` within that parameterization.
+
+The reason is that these don't change the semantics in any way -- they are
+constructively the same results.
+
+This should be the initial implementation _goal_, not a requirement for any
+incremental progress on implementation. Short-circuiting or other approximations
+used in incrementally building towards this goal should always be reasonable to
+consider based on the tradeoffs at hand. But they should be seen as temporary,
+incremental measures and where possible removed so we can evaluate the full
+semantic model when the implementation is complete.
+
+Based on the practical costs that this kind of model incurs once we have it
+implemented, we should first look for ways to adjust the design itself to avoid
+forming the complex semantic expansion in enough cases to make the model
+scalable. And only when we've exhausted those options begin adding long-term
+shortcuts in the SemIR model for common cases with a option to disable them for
+testing and debugging.
+
+## Rationale
```

Comment:
One thing I think would be worth discussing here is the impact of this proposal on cycles in the design: it'd be easy for us to accidentally define (for example) binding in terms of function calls and function calls in terms of bindings in such a way that we'd never bottom out. By requiring the toolchain to follow the design and not take shortcuts, at least under a flag, we can empirically test that our foundations actually work.

This means, for example, however we cut the loop of a function call desugaring into another function call needs to be determined in the design, not merely done in the toolchain at a convenient spot. So we shouldn't have the toolchain say "this is a direct call to a function; I'm not using the full function call machinery to dispatch it, because that would end up resulting in another function call", and instead we should have the *design* say that, if that's the intended approach. And if the intended approach is instead that we do use the full function call machinery, including `impl` lookup, but only *once*, then the design should say that instead.

Reply:
I tried adding a section about this. Not sure I really explained it well though, so ssuggestions welcome. =]